plugins/circleci/src/hooks/usePipelines.ts

```diff
@@ -0,0 +1,137 @@
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+import { useCallback, useEffect, useRef, useState } from 'react';
+import { Pipeline, PipelineInfo, Workflow } from '../types';
+import { errorApiRef, useApi } from '@backstage/core-plugin-api';
+import { AuthenticationError } from '@backstage/errors';
+import useAsyncRetry from 'react-use/lib/useAsyncRetry';
+import { useProjectSlugFromEntity } from './useProjectSlugFromEntity';
+import { circleCIApiRef } from '../api';
+
+type PipelinesPage = {
+  items?: PipelineInfo[];
+  pageToken?: string;
+  nextPageToken?: string;
+};
+
+export function usePipelines() {
+  const { repo, owner, projectSlug } = useProjectSlugFromEntity();
+  const api = useApi(circleCIApiRef);
+  const errorApi = useApi(errorApiRef);
+  const pipelinesPage = useRef<PipelinesPage>({ items: [] });
+  const [pageToken, setPageToken] = useState<string>();
+  const [pipelines, setPipelines] = useState<PipelineInfo[]>([]);
+
+  const getWorkflows = useCallback(
+    async (pipelineId: string) => {
+      try {
+        const response = await api.getWorkflowsForPipeline(pipelineId);
+        return response.items;
+      } catch (e) {
+        if (!(e instanceof AuthenticationError)) {
+          errorApi.post(e);
+        }
+        return Promise.reject(e);
+      }
+    },
+    [api, errorApi],
+  );
+
+  const transform = async (
+    pipelinesData: Pipeline[],
+    getWorkflowsCall: { (pipelineId: string): Promise<Workflow[]> },
+  ): Promise<PipelineInfo[]> => {
+    const pipelinePromises = pipelinesData.map(async pipelineData => {
+      const workflows = await getWorkflowsCall(pipelineData.id);
+      const tableBuildInfo: PipelineInfo = {
+        ...pipelineData,
+        workflows,
+      };
+      return tableBuildInfo;
+    });
+
+    return Promise.all(pipelinePromises);
+  };
+
+  const getPipelines = useCallback(async (): Promise<PipelineInfo[]> => {
+    if (!projectSlug) {
+      return Promise.reject(new Error('Missing project slug'));
+    }
+
+    const { items = [] } = pipelinesPage.current;
+
+    try {
+      const response = await api.getPipelinesForProject(projectSlug, pageToken);
+
+      const allPipelines = [
+        ...items,
+        ...(await transform(response.items ?? [], getWorkflows)),
+      ];
+
+      pipelinesPage.current = {
+        items: allPipelines,
+        nextPageToken: response.next_page_token,
+      };
+
+      return allPipelines;
+    } catch (e) {
+      if (!(e instanceof AuthenticationError)) {
+        errorApi.post(e);
+      }
+      return Promise.reject(e);
+    }
+  }, [projectSlug, pageToken, getWorkflows, api, errorApi]);
+
+  const { value, loading, retry } = useAsyncRetry(() => {
+    return getPipelines();
+  }, [getPipelines]);
+
+  useEffect(() => setPipelines(value || []), [value]);
+
+  const rerunWorkflow = async (workflowId: string) => {
```

Comment:
You may want to make these callbacks wrapped in `useCallback` too, so that your return values are stable to the caller. So if the caller in turn has `rerunWorkflow` as a dependency in their own stuff, it won't lead to infinite repaints.

This goes for the others below too of course

Reply:
Fixed